## Homework 1, CS685 Fall 2021

### This is due on November 5th, 2021. This notebook is to be submitted via Gradescope as a PDF file, while your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) should be emailed to cs685instructors@gmail.com with the subject line formatted as **Firstname_Lastname_HW1data**. 100 points total.

#### IMPORTANT: After copying this notebook to your Google Drive, please paste a link to it below. To get a publicly-accessible link, hit the *Share* button at the top right, then click "Get shareable link" and copy over the result. If you fail to do this, you will receive no credit for this homework!


---


##### *How to submit this problem set:*
- Write all the answers in this Colab notebook. Once you are finished, generate a PDF via (File -> Print -> Save as PDF) and upload it to Gradescope.
  
- **Important:** check your PDF before you submit to Gradescope to make sure it exported correctly. If Colab gets confused about your syntax, it will sometimes terminate the PDF creation routine early.

- **Important:** on Gradescope, please make sure that you tag each page with the corresponding question(s). This makes it significantly easier for our graders to grade submissions, especially with the long outputs of many of these cells. We will take off points for submissions that are not tagged.

- When creating your final version of the PDF to hand in, please do a fresh restart and execute every cell in order. One handy way to do this is by clicking `Runtime -> Run All` in the notebook menu.

---

##### *Academic honesty*

- We will audit the Colab notebooks from a set number of students, chosen at random. The audits will check that the code you wrote actually generates the answers in your PDF. If you turn in correct answers on your PDF without code that actually generates those answers, we will consider this a serious case of cheating. See the course page for honesty policies.

- We will also run automatic checks of Colab notebooks for plagiarism. Copying code from others is also considered a serious case of cheating.

---

# Part 1: Annotation

In this homework, you will first collect a labeled dataset of **120** sentences for a text classification task of your choice. This process will include:

1. *Data collection*: Collect 120 sentences from any source you find interesting (e.g., literature, Tweets, news articles, reviews, etc.)

2. *Task design*: Come up with a binary (i.e., only two labels) sentence-level classification task that you would like to perform on your sentences. Don't choose something boring like sentiment analysis... Be creative! Write up annotator guidelines/instructions on how you would like people to label your data.

3. On your dataset, collect annotations from **two** classmates for your task. Everyone in this class will need to both create their own dataset and also serve as an annotator for two other classmates. In order to get everything done on time, you need to complete the following steps:

> *   Find two classmates willing to label 120 sentences each (we will open a Piazza thread to help facilitate this).
*   Send them your annotation guidelines and a way that they can easily annotate the data (e.g., a spreadsheet or Google form)
*   Collect the labeled data from each of the two annotators.
*   Sanity check the data for basic cleanliness (are all examples annotated? are all labels allowable ones?)

4. Collect feedback from annotators about the task including annotation time and obstacles encountered (e.g., maybe your guidelines were confusing! or maybe some sentences were particularly hard to annotate!)

5. Calculate and report inter-annotator agreement.

6. Aggregate output from both annotators to create final dataset.

7. Perform NLP experiments on your new dataset!

## Question 1.1 (10 points):
Describe the source of your unlabeled data, why you chose it, and what kind of sentence selection process you used (if any) to choose 120 sentences for annotation. Also briefly describe the text classification task that you will be collecting labels for in the next section.

####*SOURCE OF DATA: PERSONAL E-MAIL ACCOUNT*####
####*WHY CHOOSE IT: The goal is to create a “Personalization Mail Assistant” for myself. Many a times, I face trouble while finding a useful mail. Thus, I am trying to create something that I mark as important/relevant for me so that I can quickly go back to these mails whenever required. Goal is also to evaluate and understand how well a transformer model will be able to perform transfer learning using such highly personalized instances unlike normal tasks like sentiment analysis, textual entailment etc.*####
####*SENTENCE SELECTION PROCESS: The sentence selection process was from mail. For relevant sentences, I decided to choose already starred mails from my previous experience and for the other class, I chose sentences from fields like social media mails, promotions, non-relevant mails. Also, while choosing the sentences, I chose close to 40% of positive class instances and 60% negative class instances so that model sees close to equal numbers in these categories but kept number for positive class a bit low to mimick real life scenario*####
####*TASK DESCRIPTION for TEXT CLASSIFICATION: The task is to create a “Personalization Mail Assistant”for myself. There are two labels: 1(Starrable/Important Mails) i.e. mail contents that might be very relevant and personal for me and 0(Non-Starrable/Non-Relevant Mails) which include categories which are not personally relevant for me. So, the idea is when a mail enters my inbox, I can use this model to immediatelely categorize into personally important vs non-important so that I can refer back to it in future quickly*####

## Question 1.2 (25 points):
Copy the annotation guidelines that you provided to your classmates below. We expect that these guidelines will be very detailed (and as such could be fairly long). You must include:

> *   The two categories for your binary classification problem, including the exact strings you want annotators to use while labeling. 
*   Descriptions of the categories and what they mean.
*   Representative examples of each category (i.e., sentences from outside your dataset that you have manually labeled to give annotators an idea of how to perform the task)
*   A discussion of of tricky corner cases, and criteria to help the annotator decide them. If you look at the data and think about how an annotator could do the task, you will likely find a bunch of these!


---

## ANNOTATION GUIDELINES##
####GENERAL TASK INTRODUCTION: The task is to create a “Personalization Mail Assistant”. Think of this as a “label:starred” feature in GMail. But having a more personal touch as I am trying to create something that I mark as important/relevant for me so that I can quickly go back to these mails whenever required. Goal is also to evaluate how well a transformer model will be to perform transfer learning using such highly personalized instances.####

####CATEGORIES: There are two labels: 1(Starrable/Relevant Mails) i.e. mail contents that might be very relevant and personal for me and 0(Non-Starrable/Non-Relevant Mails).####

####CATEGORY DESCRIPTION: ####
##### FOR 1(Starrable/Relevant Mails): 
*   Information that is personal in nature for a student or that of a prospective employee are starrable and important (like offer letter, online assessment information, interview information, rejection information, acceptance information)
*   Information that mention about an important attachment in the mail, a zoom/meeting link in the mail, card information fall in this category
* Information from colleges, CICS, conferences and publications and other similar institutions are starrable
* Important registrations like flights, conferences etc.
* Information between faculty and students are starrable


##### FOR 0(Non-Starrable/Non-Relevant Mails): 
*   Promotional Mails having offers are non-starrable (like 10% discount, visit shop, 300 off etc.)
*   Abstract sentences like quotes are definitely non-starrable
*   Notifications from social media like (birthday and friend/LinkedIn request) are non-starrable 
*  Random sentences that have no significance for a student or feels like a part of a paragraph from somewhere also falls into this category (e.g. Having a detailed financial & business plan before approaching investors helps entrepreneurs in getting the right kind of funding.)

#### EXAMPLES####
##### FOR 1(Starrable/Relevant Mails): 
*   We will let you know when our 2022 roles start to be posted in the coming month or so to our University Careers Page. We look forward to seeing your applications.
*  We are asking for your help in holding additional office hours and answering piazza posts, if that fits within your time commitment.
The students have asked many questions on piazza and the TAs are quite overwhelmed.
* Your flight ticket for Bangalore-Patna-Bangalore is confirmed

##### FOR 0(Non-Starrable/Non-Relevant Mails): 
*   Whether you’re decorating, shopping for a gift, or looking to replace that faulty blender, everything you need from Bed Bath & Beyond is now easily available on DoorDash.
*   Whether you’re decorating, shopping for a gift, or looking to replace that faulty blender, everything you need from Bed Bath & Beyond is now easily available on DoorDash.
* Ends Friday: 30% off your pass to reconnecting with the data science community


#### CORNER CASES: In cases of confusion, I would suggest to re-read the sentence and then ask yourself a question: as a student or an employee do you think this information can be relevant at personal level. Also, if it is promotional or social media related content, feel free to directly flag it as non-starrable (0). I have added one more Column: “Not-Confident”. In case still, you feel that you are not confident enough, select the best category and set this flag as 1. I will leverage this flag for conflict resolution or determining the best label from the two labellers. 

## Question 1.3 (5 points):
Write down the names and emails of the two classmates who will be annotating your data below. Once they are finished annotating, create two .csv files (annotator1.csv and annotator2.csv) that contains each annotator's labels. The file should have two columns with headers **text** and **label**, respectively. You will include these files in an email to the instructors account when you're finished with this homework. 

*The tweets.csv file provided as an example in Part 2 below uses the same format.*

### *Annotator 1: Abhishek Lalwani (alalwani@umass.edu)* ###
### *Annotator 2: Prahlad Das (prahladdas@umass.edu)* ###

## Question 1.4 (10 points):
After both annotators have finished labeling the 120 sentences you gave them, ask them for feedback about your task and the provided annotation guidelines. If you were to collect more labeled data for this task in the future, what would you change from your current setup? Why? Please include a summary of annotator feedback (with specific examples that they found challenging to label) in your answer.


### What would you change from current setup and why ? ###

*   If I were to change something in the current setup, I would reselect those examples which I feel are harder for model to classify correctly by the model. Reasoning: This way we can ensure that the model actually learns the cases where the differentiation is not very significant (instances close to the decision boundary) instead of just classifying simple instances. 

### Feedback Summary From Prahlad Das ###
*   It was good to annotate the data as I could relate all of them with my emails. Though there were a few tricky cases where it was difficult to decide like ' We connect you with 3 volunteers and 4 peers who will help you find internships.' and ' Covering topics such as LSTMs, generative adversarial transformers, embeddings, and more, this is some standout NLP research from 2021 so far. ' I was not sure if they were spam or requested services as no other information was available.
Annotation guideline was clear with description of both categories with nice examples. Even how to resolve corner cases was explained clearly.

### Feedback Summary from Abhishek Lalwani ###
*   Annotation guidelines were very clear and most of the selected examples were meticulously chosen. Only in some cases, I felt that it would have been better if there were more than two categories for annotation but given binary nature of the task, it was understandable. One of the example sentence that I found confusing was: "Covering topics such as LSTMs, generative adversarial transformers, embeddings, and more, this is some standout NLP research from 2021 so far." as it was difficult to determine which class this belonged to without much context




## Question 1.5 (10 points):
Now, compute the inter-annotator agreement between your two annotators. Upload both .csv files to your Colab session (click the folder icon in the sidebar to the left of the screen). In the code cell below, read the data from the two files and compute both the raw agreement (% of examples for which both annotators agreed on the label) and the [Cohen's Kappa](https://en.wikipedia.org/wiki/Cohen%27s_kappa). Feel free to use implementations in existing libraries (e.g., [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html)). After you're done, paste the numbers in the text cell that follows your code. 

*If you're curious, Cohen suggested the Kappa result be interpreted as follows: values ≤ 0 as indicating no agreement and 0.01–0.20 as none to slight, 0.21–0.40 as fair, 0.41– 0.60 as moderate, 0.61–0.80 as substantial, and 0.81–1.00 as almost perfect agreement.*

In [ ]:
### WRITE CODE TO LOAD ANNOTATIONS AND 
import pandas as pd
import numpy as np
from sklearn.metrics import cohen_kappa_score
data1 = pd.read_csv("annotator1.csv")
data2 = pd.read_csv("annotator2.csv")
label1 = data1['label'].tolist()
label2 = data2['label'].tolist()
data1['label'] = data1['label'].astype('int32')
data2['label'] = data2['label'].astype('int32')
### COMPUTE AGREEMENT + COHEN'S KAPPA HERE!
suma = 0
for i in range(len(label1)):
  if label1[i]==label2[i]:
    suma+=1
raw_agreement = suma*100/len(label1)
ck_score      = cohen_kappa_score(data1['label'],data2['label'])
print("Raw Agreement is: ", raw_agreement)
print("Cohen Kappa Score is: ", ck_score)

Raw Agreement is:  90.0
Cohen Kappa Score is:  0.7869192068659367


###*RAW AGREEMENT*: 90.0%
###*COHEN'S KAPPA*: 0.786

## Question 1.6 (10 points):
To form your final dataset, you need to *aggregate* the annotations from both annotators (i.e., for cases where they disagree, you need to choose a single label). Use any method you like other than random label selection to perform this aggregation (e.g., have the two annotators discuss each disagreement and come to consensus, or choose the label you agree with the most). Upload your final dataset to the Colab session (in the same format as the other two files) as final_dataset.csv. Remember to include this file in your final email to us!

### *I had asked where the two annotators were not confident in the annotation sheet that they had shared with me. First I used that for the resolution and selected the one that I thought to be best with my  and more confident annotator's agreement. For other remaining ones, I chose the one that I had already starred in my e-mail as they were the true labels in my case (i.e. my perosonal best agreement)* ###

# Part 2: Text classification

Now we'll move onto fine-tuning  pretrained language models specifically on your dataset. This part of the homework is meant to be an introduction to the HuggingFace library, and it contains code that will potentially be useful for your final projects. Since we're dealing with large models, the first step is to change to a GPU runtime.

## Adding a hardware accelerator

Please go to the menu and add a GPU as follows:

`Edit > Notebook Settings > Hardware accelerator > (GPU)`

Run the following cell to confirm that the GPU is detected.

In [ ]:
import torch

# Confirm that the GPU is detected

assert torch.cuda.is_available()

# Get the GPU device name.
device_name = torch.cuda.get_device_name()
n_gpu = torch.cuda.device_count()
print(f"Found device: {device_name}, n_gpu: {n_gpu}")
device = torch.device("cuda")

Found device: Tesla K80, n_gpu: 1


## Installing Hugging Face's Transformers library
We will use Hugging Face's Transformers (https://github.com/huggingface/transformers), an open-source library that provides general-purpose architectures for natural language understanding and generation with a collection of various pretrained models made by the NLP community. This library will allow us to easily use pretrained models like `BERT` and perform experiments on top of them. We can use these models to solve downstream target tasks, such as text classification, question answering, and sequence labeling.

Run the following cell to install Hugging Face's Transformers library and download a sample data file called tweets.csv that contains tweets about airlines along with a negative, neutral, or positive sentiment rating. Note that you will be asked to link with your Google Drive account to download some of these files. If you're concerned about security risks (there have not been any issues in previous semesters), feel free to make a new Google account and use it for this homework!

In [ ]:
!pip install transformers
!pip install -U -q PyDrive

from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
print('success!')

import os
import zipfile

# Download helper functions file
helper_file = drive.CreateFile({'id': '16HW-z9Y1tM3gZ_vFpJAuwUDohz91Aac-'})
helper_file.GetContentFile('helpers.py')
print('helper file downloaded! (helpers.py)')

# Download sample file of tweets
data_file = drive.CreateFile({'id': '1QcoAmjOYRtsMX7njjQTYooIbJHPc6Ese'})
data_file.GetContentFile('tweets.csv')
print('sample tweets downloaded! (tweets.csv)')

success!
helper file downloaded! (helpers.py)
sample tweets downloaded! (tweets.csv)


The cell below imports some helper functions we wrote to demonstrate the task on the sample tweet dataset.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy

# Part 1: Data Prep and Model Specifications

Upload your data using the file explorer to the left. We have provided a function below to tokenize and format your data as BERT requires. Make sure that your csv file, titled final_data.csv, has one column "text" and another column "labels" containing integers.

If you run the cell below without modifications, it will run on the tweets.csv example data we have provided. It imports some helper functions we wrote to demonstrate the task on the sample tweet dataset. You should first run all of the following cells with tweets.csv just to see how everything works. Then, once you understand the whole preprocessing / fine-tuning process, change the csv in the below cell to your final_data.csv file, add any extra preprocessing code you wish, and then run the cells again on your own data.

In [ ]:
from helpers import tokenize_and_format, flat_accuracy
import pandas as pd

df = pd.read_csv('final_data.csv')
#df = pd.read_csv('tweets.csv')

df = df.sample(frac=1).reset_index(drop=True)

texts = df.text.values
labels = df.label.values

### tokenize_and_format() is a helper function provided in helpers.py ###
input_ids, attention_masks = tokenize_and_format(texts)

# Convert the lists into tensors.
input_ids = torch.cat(input_ids, dim=0)
attention_masks = torch.cat(attention_masks, dim=0)
labels = torch.tensor(labels)

# Print sentence 0, now as a list of IDs.
print('Original: ', texts[0])
print('Token IDs:', input_ids[0])

Original:  As a result of this particular day, we welcomed approximately 40,000 new retail colleagues, including pharmacists, store associates, and pharmacy technicians!
Token IDs: tensor([  101,  2004,  1037,  2765,  1997,  2023,  3327,  2154,  1010,  2057,
        10979,  3155,  2871,  1010,  2199,  2047,  7027,  8628,  1010,  2164,
         6887, 27292,  6305,  5130,  1010,  3573,  9228,  1010,  1998, 13882,
        20202,   999,   102,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0])


## Create train/test/validation splits

Here we split your dataset into 3 parts: a training set, a validation set, and a testing set. Each item in your dataset will be a 3-tuple containing an input_id tensor, an attention_mask tensor, and a label tensor.



In [ ]:
total = len(df)

num_train = int(total * .8)
num_val = int(total * .1)
num_test = total - num_train - num_val

# make lists of 3-tuples (already shuffled the dataframe in cell above)

train_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train)]
val_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_train, num_val+num_train)]
test_set = [(input_ids[i], attention_masks[i], labels[i]) for i in range(num_val + num_train, total)]

train_text = [texts[i] for i in range(num_train)]
val_text = [texts[i] for i in range(num_train, num_val+num_train)]
test_text = [texts[i] for i in range(num_val + num_train, total)]

Here we choose the model we want to finetune from https://huggingface.co/transformers/pretrained_models.html. Because the task requires us to label sentences, we wil be using BertForSequenceClassification below. You may see a warning that states that `some weights of the model checkpoint at [model name] were not used when initializing. . .` This warning is expected and means that you should fine-tune your pre-trained model before using it on your downstream task. See [here](https://github.com/huggingface/transformers/issues/5421#issuecomment-652582854) for more info.

In [ ]:
from transformers import BertForSequenceClassification, AdamW, BertConfig

model = BertForSequenceClassification.from_pretrained(
    "bert-base-uncased", # Use the 12-layer BERT model, with an uncased vocab.
    num_labels = 2, # The number of output labels.   
    output_attentions = False, # Whether the model returns attentions weights.
    output_hidden_states = False, # Whether the model returns all hidden-states.
)

# Tell pytorch to run this model on the GPU.
model.cuda()

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForSequenceClassification: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.weight', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at

BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0): BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, element

# ACTION REQUIRED #

Define your fine-tuning hyperparameters in the cell below (we have randomly picked some values to start with). We want you to experiment with different configurations to find the one that works best (i.e., highest accuracy) on your validation set. Feel free to also change pretrained models to others available in the HuggingFace library (you'll have to modify the cell above to do this). You might find papers on BERT fine-tuning stability (e.g., [Mosbach et al., ICLR 2021](https://openreview.net/pdf?id=nzpLWnVAyah)) to be of interest.

In [ ]:
batch_size = 32
optimizer = AdamW(model.parameters(),
                  lr = 8e-6, # args.learning_rate - default is 5e-5
                  eps = 1e-7 # args.adam_epsilon  - default is 1e-8
                )
epochs = 20

# Fine-tune your model
Here we provide code for fine-tuning your model, monitoring the loss, and checking your validation accuracy. Rerun both of the below cells when you change your hyperparameters above.

In [ ]:
import numpy as np
# function to get validation accuracy
def get_validation_performance(val_set):
    # Put the model in evaluation mode
    model.eval()

    # Tracking variables 
    total_eval_accuracy = 0
    total_eval_loss = 0

    num_batches = int(len(val_set)/batch_size) + 1

    total_correct = 0

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]
      
      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])
      
      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)
        
      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():        

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids, 
                                token_type_ids=None, 
                                attention_mask=b_input_mask,
                                labels=b_labels)
        loss = outputs.loss
        logits = outputs.logits
            
        # Accumulate the validation loss.
        total_eval_loss += loss.item()
        
        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()
        num_correct = np.sum(pred_flat == labels_flat)
        total_correct += num_correct
        
    # Report the final accuracy for this validation run.
    avg_val_accuracy = total_correct / len(val_set)
    return avg_val_accuracy

In [ ]:
import random

# training loop

# For each epoch...
for epoch_i in range(0, epochs):
    # Perform one full pass over the training set.

    print("")
    print('======== Epoch {:} / {:} ========'.format(epoch_i + 1, epochs))
    print('Training...')

    # Reset the total loss for this epoch.
    total_train_loss = 0

    # Put the model into training mode.
    model.train()

    # For each batch of training data...
    num_batches = int(len(train_set)/batch_size) + 1

    for i in range(num_batches):
      end_index = min(batch_size * (i+1), len(train_set))

      batch = train_set[i*batch_size:end_index]

      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])

      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)

      # Clear the previously calculated gradient
      model.zero_grad()        

      # Perform a forward pass (evaluate the model on this training batch).
      outputs = model(b_input_ids, 
                            token_type_ids=None, 
                            attention_mask=b_input_mask, 
                            labels=b_labels)
      loss = outputs.loss
      logits = outputs.logits

      total_train_loss += loss.item()

      # Perform a backward pass to calculate the gradients.
      loss.backward()

      # Update parameters and take a step using the computed gradient.
      optimizer.step()
        
    # ========================================
    #               Validation
    # ========================================
    # After the completion of each training epoch, measure our performance on
    # our validation set. Implement this function in the cell above.
    print(f"Total loss: {total_train_loss}")
    val_acc = get_validation_performance(val_set)
    print(f"Validation accuracy: {val_acc}")
    
print("")
print("Training complete!")



======== Epoch 1 / 20 ========
Training...
Total loss: 1.895131528377533
Validation accuracy: 0.5833333333333334

======== Epoch 2 / 20 ========
Training...
Total loss: 1.7595832347869873
Validation accuracy: 0.6666666666666666

======== Epoch 3 / 20 ========
Training...
Total loss: 1.6575713753700256
Validation accuracy: 0.8333333333333334

======== Epoch 4 / 20 ========
Training...
Total loss: 1.4999759793281555
Validation accuracy: 0.8333333333333334

======== Epoch 5 / 20 ========
Training...
Total loss: 1.294845700263977
Validation accuracy: 0.8333333333333334

======== Epoch 6 / 20 ========
Training...
Total loss: 1.1287816166877747
Validation accuracy: 0.8333333333333334

======== Epoch 7 / 20 ========
Training...
Total loss: 0.9871845245361328
Validation accuracy: 0.9166666666666666

======== Epoch 8 / 20 ========
Training...
Total loss: 0.8789694011211395
Validation accuracy: 0.9166666666666666

======== Epoch 9 / 20 ========
Training...
Total loss: 0.7518913596868515
Validat

# Evaluate your model on the test set
After you're satisfied with your hyperparameters (i.e., you're unable to achieve higher validation accuracy by modifying them further), it's time to evaluate your model on the test set! Run the below cell to compute test set accuracy.


In [ ]:
get_validation_performance(test_set)

0.9166666666666666

## Question 2.1 (10 points):
Congratulations! You've now gone through the entire fine-tuning process and created a model for your downstream task. Two more questions left :) First, describe your hyperparameter selection process in words. If you based your process on any research papers or websites, please reference them. Why do you think the hyperparameters you ended up choosing worked better than others? Also, is there a significant discrepancy between your test and validation accuracy? Why do you think this is the case?

### 


*   *Hyperparameter Selection Process: First I ran a baseline model with very low learning rate of order 1e-8 with good number of epochs but since the learning rate was low initially, the validation accuracy was stuck at 0.583. So, I decided to ramp up the learning rate but referred a paper before that. In the paper (https://openreview.net/pdf?id=nzpLWnVAyah and https://arxiv.org/pdf/1905.05583.pdf) authors highlight the significance of choosing a correct learning rate and batch size. I decided  to choose the same batch size of 20 mentioned in this paper and a learning rate in a similar range but a bit smaller (8e-6 instead of 2e-5 mentioned in the paper).*
*   *Why hyperparameters worked better than others ? --> Learning rate is very important. Initially extremely small learning rate might lead to a slow convergence and the accuracy might get stuck in the local minima as the step size would be small. A very high learning rate in the finetuning process leads to the problem of catastrophic forgetting and hence a learning rate in the range of [7e-6, 5e-5] works very good in my case. Also, based on experiments, results from higher batch sizes are better. In our case, 20 is selected based on paper above. We could have selected 32 as mentioned in Devlin et. al. (arXiv preprint arXiv:1810.04805) but since we see that in our case, network starts converging around 15 batch size, we allow to train a bit more and stop at 20 batch size.*
* *Also, is there a significant discrepancy between your test and validation accuracy?* : No
* *Why do you think this is the case?* In our case, we can see that the validation accuracy is same as that of the test accuracy which leads us to believe that there is no overfitting which is true to a significant extent but there are certain things that must be kept in mind before using this model in actual scenarios: The vallidation and test examples are very small and hence the variance of the data could be high in some cases and thus a better strategy would be to create a bigger dataset so that we have significant number of validation and test examples with us. Also, since this validation and test data is small, variability could also be observed based on what examples are sampled at the starting of the notebook and even 1 misclassification can result in drastic shifts (~10% as our validation size is only 12). So, we should always keep these factors in our mind. 




## Question 2.2 (20 points):
Finally, perform an *error analysis* on your model. This is good practice for your final project. Write some code in the below code cell to print out the text of up to five test set examples that your model gets **wrong** (if your model gets more than five test examples wrong, randomly choose five of them to print out). Then, in the following text cell, perform a qualitative analysis of these examples. See if you can figure out any reasons for errors that you observe, or if you have any informed guesses (e.g., common linguistic properties of these particular examples). Does this suggest any possible future steps could for your classifier?

In [ ]:
## YOUR ERROR ANALYSIS CODE HERE
## print out up to 5 test set examples that your model gets wrong

In [ ]:
import numpy as np
def get_valpred_labels(val_set):
    # Put the model in evaluation mode
    model.eval()

    # Tracking variables 
    total_eval_accuracy = 0
    total_eval_loss = 0

    num_batches = int(len(val_set)/batch_size) + 1

    total_correct = 0

    for i in range(num_batches):

      end_index = min(batch_size * (i+1), len(val_set))

      batch = val_set[i*batch_size:end_index]
      
      if len(batch) == 0: continue

      input_id_tensors = torch.stack([data[0] for data in batch])
      input_mask_tensors = torch.stack([data[1] for data in batch])
      label_tensors = torch.stack([data[2] for data in batch])
      
      # Move tensors to the GPU
      b_input_ids = input_id_tensors.to(device)
      b_input_mask = input_mask_tensors.to(device)
      b_labels = label_tensors.to(device)
        
      # Tell pytorch not to bother with constructing the compute graph during
      # the forward pass, since this is only needed for backprop (training).
      with torch.no_grad():        

        # Forward pass, calculate logit predictions.
        outputs = model(b_input_ids, 
                                token_type_ids=None, 
                                attention_mask=b_input_mask,
                                labels=b_labels)
        loss = outputs.loss
        logits = outputs.logits
            
        # Accumulate the validation loss.
        total_eval_loss += loss.item()
        
        # Move logits and labels to CPU
        logits = logits.detach().cpu().numpy()
        label_ids = b_labels.to('cpu').numpy()

        # Calculate the number of correctly labeled examples in batch
        pred_flat = np.argmax(logits, axis=1).flatten()
        labels_flat = label_ids.flatten()
        num_correct = np.sum(pred_flat == labels_flat)
        total_correct += num_correct
        return labels_flat, pred_flat

In [ ]:
labels_actual, pred_labels = get_valpred_labels(test_set)
index_list = [] #Find index of non-matching labels and append it in this
for i in range(len(labels_actual)):
  if labels_actual[i]!=pred_labels[i]:
    index_list.append(i)
print("Non-Matching test set predictions are: ")
for i in range(len(index_list)):
  print(" ")
  print("Incorrect Sentence #"+ str(i+1) + " is: ")
  print(" ")
  print(test_text[i])

Non-Matching test set predictions are: 
 
Incorrect Sentence #1 is: 
 
No one would have crossed the ocean if he could have gotten off the ship in the storm


### *DESCRIBE YOUR QUALITATIVE ANALYSIS OF THE ABOVE EXAMPLES HERE*

In my case, only one of the example is incorrect. Thus based on Mohit's suggestion in Piazza post, I am explaining the possible sets of reasons below. The sentence is "No one would have crossed the ocean if he could have gotten off the ship in the storm". We see in this example, there is mention of pronoun "he" which gives it a personal touch and might have confused the model to some extent but since there is only one incorrect example, the actual reasoning is very hard to ascertain. There could be multiple reasons as to why our model is performing well and I detail them in the upcoming pointers:

*   If we see annotation guideline, the sentence is not promotional, neither comes from social media, nor is abstract and rather has a slight personal touch. Though for a human based on annotation guideline no 4, it is clearly not for students and a random sentence. While it is easy for a human annotator, it might have been difficult for the model to comprehend and classify accurately. Still, the model does extremely well on all the other examples. 
*   Since, we are testing only on 12 examples, we can say that we might have little confidence in our results and results would slightly also depend on what we sampled as a part of train, test and validate split.
* Another, important reason is how we select the data also influences the entire accuracy of the pipeline. It is important that we choose tough examples for the model so that it is able to classify correctly even in tough cases. Model might have caught on to certain keywords and hence it became easier to classify two classes easily. Hence, present selection of validation and testing examples are way too easy for the model to classify incorrectly. ( Or probably how I filter mails in my actual life i.e. annotation guidelines are way too sorted :) )
* Present selection of model and its hyperparameters are very good. We see that at extremely low learning rate the accuracy falls to 58.3%. Also, we know that due to the multiheaded self-attention mechanism, BERT is able to create very good contextual embedding and thus in this case able to generate a very good decision boundary which leads to a very good classification performance.  



---

Finished? Remember to upload the PDF file of this notebook to Gradescope **AND** email your three dataset files (annotator1.csv, annotator2.csv, and final_data.csv) to cs685instructors@gmail.com with the subject line formatted as **Firstname_Lastname_HW1data**.
